File: class23_homework/A-Ivaylo-Markov-12/java/src/org/elsys/se/class23/homework/plane/Plane.java
package org.elsys.se.class23.homework.plane;

import org.elsys.se.class23.homework.people.*;

public class Plane {
	private Passenger[][] leftSeats;
	private Passenger[][] rightSeats;
	private int seatRows;
	private int seatColumns;
	
	public Plane(int rows, int cols) {
		seatRows = rows;
		seatColumns = cols; 
		leftSeats = new Passenger[seatRows][seatColumns/2];
		rightSeats = new Passenger[seatRows][seatColumns/2];
	}
	
	public boolean addGroup(Group group) {
		boolean seatedLeft = true;
		int[] firstSeat = new int[2]; //row,column
		firstSeat = findEmptySeats(leftSeats,group.getSize());
		if(firstSeat[0] == -1 && firstSeat[1] == -1 ) {
			seatedLeft = false;
			firstSeat = findEmptySeats(rightSeats,group.getSize());
			if(firstSeat[0] == -1 && firstSeat[1] == -1 ) {
				return false;
			}
		}
		if(seatedLeft == true) {
			seatGroup(group,firstSeat,leftSeats);
		}
		else {
			seatGroup(group,firstSeat,rightSeats);
		}
		return true;
	}
	
	private void seatGroup(Group group, int[] seatNum, Passenger[][] seats) {
		for(int person=0;person<group.getSize();++person) {
			seats[seatNum[0]][seatNum[1]] = group.getMember(person);
			++seatNum[1];

		}
	}
	
	private int[] findEmptySeats(Passenger[][] seats, int amount) {
		int[] seatNumber = {-1,-1}; //row,column
		int free = 0, seat = 0;
		for(int row=0;row<seatRows;++row) {
			free = 0;
			seat = 0;
			for(int col=0;col<seatColumns/2;++col) {
				if(seats[row][col] == null) {
					if(free==0)seat = col;
					++free;
					if(free == amount) {
						seatNumber[0] = row;
						seatNumber[1] = seat;
						free = 0;
						seat = 0;
						return seatNumber;
					}
				}
				else {
					if(free!=0) {
						free = 0;
						seat = 0;
						break;
					}
				}
			}
		}
		return seatNumber;
	}
	
	public void fly() {
		System.out.println("PIKIRAM");
	}
	
	public void printSeatsPlan() {
		System.out.println("Left seats:");
		printSeats(leftSeats);
		System.out.println("Right seats:");
		printSeats(rightSeats);
		System.out.println();
	}
	
	private void printSeats(Passenger[][] seats) {
		for(int row=0;row<seatRows;++row) {
			for(int col=0;col<seatColumns/2;++col) {
				if(seats[row][col] == null) {
					System.out.print("[ ]");
				}
				else {
					System.out.print("[O]");
				}
			}
			System.out.println();
		}
	}

	
	
}

File: class23_homework/A-Ivaylo-Markov-12/ruby/plane.rb
class Plane
	
	def initialize rows,cols
		@left_seats = Array.new(rows){Array.new(cols/2)}
		@right_seats = Array.new(rows){Array.new(cols/2)}
		@availiable_row = 0
		@availiable_col = 0
		@availiable = false
	end
	
	def fly
		print "PIKIRAM"
	end
	
	def print_seats_plan
		puts "Left seats:"
		print_seats @left_seats
		puts "Right seats:"
		print_seats @right_seats
	end

	def add_group group
		seatedLeft = true
		find_empty_seats @left_seats, group.get_size
		if(!@availiable)
			seatedLeft = false
			find_empty_seats @right_seats, group.get_size
			if(!@availiable) then
				puts "Can't accommodate group of #{group.get_size}"
				return false 
			end
		end
		if(seatedLeft)
			seat_group group, @left_seats
			return true
		end
		seat_group group, @right_seats
		return true
	end

	private
	
	def print_seats seats
		seats.each_index do |row|
			seats[row].each_index do |col|
				print "[ ]" if(seats[row][col] == nil)
				print "[O]" if (seats[row][col] != nil)
			end
			puts ""
		end
	end
	
	def seat_group group, seats
		i = 0
		while (i<group.get_size) do
			seats[@availiable_row][@availiable_col] = group.get_member i
			@availiable_col +=1
			i+=1
		end
		@availiable = false
		puts "Group of #{group.get_size} accommodated."
	end
	
	def find_empty_seats seats, amount
		@availiable = false
		if(amount == 1) then
			seats.each_index do |row|
				seats[row].each_index do |col|
					if(seats[row][col] == nil) then
						@availiable = true
						@availiable_col = col
						@availiable_row = row
					end
				end
			end
		end
		
		if(amount == 2) then
			seats.each_index do |row|
				if((seats[row][0] == nil) && (seats[row][1] == nil)) then
						@availiable = true
						@availiable_col = 0
						@availiable_row = row
				end
				if((seats[row][1] == nil) && (seats[row][2] == nil)) then
						@availiable = true
						@availiable_col = 1
						@availiable_row = row
				end
			end
		end
		
		if(amount == 3) then
			seats.each_index do |row|
				if((seats[row][0] == nil) && (seats[row][1] == nil) && (seats[row][2] == nil)) then
					@availiable = true
					@availiable_col = 0
					@availiable_row = row
				end
			end
		end
		
	end
	
end

